src/plugins/intel_gpu/src/graph/impls/ocl/scaled_dot_product_attention.cpp

```diff
@@ -389,18 +414,47 @@ struct scaled_dot_product_attention_impl : multi_stage_primitive<scaled_dot_prod
         return params;
     }
 
+    static kernel_impl_params static_canonicalize_shapes(const kernel_impl_params& impl_params) {
+        auto updated_impl_params = impl_params;
+
+        auto extend_pshape_to_rank_in_num_heads_dim = [](ov::PartialShape pshape, size_t rank = 4) {
+            if (pshape.size() == rank) {
+                return pshape;
+            }
+            pshape.insert(pshape.begin() + 1, ov::Dimension(1));
+            return pshape;
+        };
+
+        // For scale of 1D tensor or attention_mask of empty shape, use extend_shape_to_rank_from_end as before
+        for (auto& input_layout : updated_impl_params.input_layouts) {
+            input_layout.set_partial_shape(input_layout.get_partial_shape().size() <= 1 ?
+                                           extend_shape_to_rank_from_end(input_layout.get_partial_shape()) :
+                                           extend_pshape_to_rank_in_num_heads_dim(input_layout.get_partial_shape()));
+        }
+
+        auto& output_layout = updated_impl_params.output_layouts[0];
+        output_layout.set_partial_shape(extend_pshape_to_rank_in_num_heads_dim(output_layout.get_partial_shape()));
+
+        return updated_impl_params;
+    }
+
+    kernel_impl_params canonicalize_shapes(const kernel_impl_params& impl_params) const override {
+        return static_canonicalize_shapes(impl_params);
+    }
+
     static std::unique_ptr<primitive_impl> create(const typed_program_node<scaled_dot_product_attention>& arg, const kernel_impl_params& impl_param) {
         std::vector<kernel_selector::kernel_data> kernels_data;
-        auto sdpa_kernel_params = get_kernel_params(impl_param, impl_param.is_dynamic());
         auto& kernel_selector = kernel_selector_t::Instance();
-        kernels_data.push_back(kernel_selector.get_best_kernel(sdpa_kernel_params));
+        auto params = static_canonicalize_shapes(impl_param);
```

Comment:
To reduce runtime copy of params, it would be better to call this function only when needed.

Reply:
What you mean is that this function should only be called in update_impl and not in impl creation?

Reply:
@andrew-k-park I think Taylor meant that we can call this static_canonicalize_shapes() function only when we have something to adjust (i.e. QKV shapes less than 4 dimensions)
We can override `update(primitive_inst& inst, const kernel_impl_params& impl_params)` function here and use either original impl_params, or apply canonicalization and use updated params depending on the QKV ranks

Reply:
yeah I just meant that canonicalizer function returnes a copy of "impl_param" and I think it is needless when the rank is 4D>

Reply:
updated as comments